psadwatchd.8
.\"
.TH PSADWATCHD 8 "March 2009" "Debian GNU/Linux"
.SH NAME
.B psadwatchd
\- checks to make sure kmsgsd and psad are running.
.SH SYNOPSIS
.B psadwatchd [options]
.SH DESCRIPTION
.B psadwatchd
checks on an interval of every five seconds to make sure that kmsgsd
and psad are running on the box.  If either of the other two daemons
have died, psadwatchd will restart the daemon and notify each
email address listed in the EMAIL_ADDRESSES variable (see
/etc/psad/psad.conf) that the daemon has been restarted.
.B psadwatchd
uses the psad.conf configuration file which by default is
located at /etc/psad/psad.conf, but a different path can be specified
on the command line.
.SH OPTIONS
.TP
.BR \-c\ \<config\-file>
Specify path to config file instead of using the default configuration file
/etc/psad/psad.conf.
.TP
.BR \-D
Dump the configuration values that
.B psadwatchd
derives from /etc/psad/psad.conf (or other override files) on STDERR.
.TP
.BR \-h
Display usage information and exit.
.TP
.BR \-O\ \<config\-file>
Override config variable values that are normally read from the
/etc/psad/psad.conf file with values from the specified file.  Multiple
override config files can be given as a comma separated list.
.SH SEE ALSO
.BR psad (8),
.BR kmsgsd (8),
.SH AUTHOR
Michael Rash (mbr@cipherdyne.org)
.PP
This manual page was written by Daniel Gubser <daniel.gubser@gutreu.ch>
for the Debian GNU/Linux system (but may be used by others).
.PP
.SH DISTRIBUTION
.B psad
is distributed under the GNU General Public License (GPL), and the latest
version may be downloaded from
.B http://www.cipherdyne.org


ChangeLog
psad-2.1.6 (04//2009):
    - (Franck Joncourt) Added --Override-config feature so that alternate
      configuration files can be specified on the command line to override
      configuration variables in the standard /etc/psad/psad.conf file.
    - (Franck Joncourt) Submitted patches to fix stderr redirection for the
      usage of the mail binary, and to close stdout, stdin, and stderr when
      running psad as a daemon.

psad-2.1.5 (02/20/2009):
    - (Miroslav Grepl) Contributed policy files to make psad compatible with
      SELinux.  The files are located in a new "selinux" directory in the
      psad sources.
    - Bug fix for local server ports not reported correctly under netstat
      parsing (Franck Joncourt).
    - (Steve B) Submitted patch to fix a bug in the start() function in the
      Gentoo init script which caused psad to not be started and the error
      "* ERROR: psad failed to start" to be generated.
    - Bug fix when ENABLE_SYSLOG_FILE is enabled to run a preliminary regex
      match on each syslog message because kmsgsd is not running and therefore
      has not gone through the kmsgsd tests for a properly structured iptables
      message.
    - Updated IPTables::Parse to 0.7.
    - Updated IPTables::ChainMgr to 0.9.

psad-2.1.4 (08/21/2008):
    - Restructured perl module paths to make it easy to introduce a "nodeps"
      distribution of psad that does not contain any perl modules.  This
      allows better integration with systems that already have all necessary
      modules installed (including the IPTables::ChainMgr and IPTables::Parse
      modules).  The main driver for this work is to make all cipherdyne.org
      projects easily integrated with distributions based on Debian, and
      Franck Joncourt has been instrumental in making this process a reality.
      All perl modules are now placed within the "deps" directory, and the
      install.pl script checks to see if this directory exists - a separate
      psad-<ver>-nodeps tarball will be distributed without this directory.
      The Debian package for psad can then reference the -nodeps tarball, and
      a new "psad-nodeps.spec" file has been added to build an RPM from the
      psad sources that does not install any perl modules.
    - Updated to use the normal system whois client if the /usr/bin/whois_psad
      path does not exist, and moved the whois/ directory into the deps/
      directory.  This removes /usr/bin/whois_psad as a strict dependency.
    - Bugfix to honor the IPT_SYSLOG_FILE variable in --Analyze-msgs mode.
    - Switched from the deprecated bleeding-all.rules file to the new
      emerging-all.rules available from Matt Jonkman at Emerging Threats
      (http://www.emergingthreats.net).

psad-2.1.3 (06/07/2008):
    - Updated to enable IPT_SYSLOG_FILE by default.  This is a relatively
      important change since it changes the method of acquiring iptables log
      data from reading it out of named pipe from syslog to just parsing the
      /var/log/messages file.  This implies that kmsgsd does not have to run,
      and that it is much easier to ensure that psad actually receives
      iptables log messages.  The most complex and error prone aspect of psad
      in the past has been the reconfiguration of the various syslog daemons
      out there (which have very different configuration syntax and features)
      to write kern.info messages to the /var/lib/psad/psadfifo named pipe.
    - Updated to version 4.7.26 of the whois client from Marco d'Itri.  This
      allows whois records for some addresses (such as 116.125.35.98, which
      which was scanning a system running psad but could not be identified
      under the older whois client) to be properly queried.
    - Updated to Bit::Vector 6.4 from 6.3.
    - Updated to Date::Calc 5.4 from 5.3.
    - Updated to Storable 2.18 from 2.16.

psad-2.1.2 (04/03/2008):
    - Bugfix to not include kernel timestamps in iptables log prefixes that
      contain spaces like "[   65.026008] DROP" (bug reported by Erik Heidt).
    - Bugfix to skip non-resolved IP addresses (bug reported by Albert Whale)
    - Better p0f output in --debug mode to display when a passive OS
      fingerprint cannot be calculated based on iptables log messages that
      include tcp options (i.e., with --log-tcp-options when building a LOG
      rule on the iptables command line).

psad-2.1.1 (01/25/2008):
    - Added a new feature whereby psad can acquire iptables log data just by
      parsing an existing file (/var/log/messages by default) that is written
      to by syslog.  By default, psad acquires iptables log data from the
      /var/log/psad/fwdata file which is written to by kmsgsd, but on some
      systems, having syslog communicate log data to kmsgsd can be problematic
      since syslog configs and external factors such as Apparmor and SELinux
      can play a role here.  This new feature is controled by two new
      configuration variables "ENABLE_SYSLOG_FILE" (to enable/disable the
      feature) and "IPT_SYSLOG_FILE" to specifiy the path to the file to
      parse.
    - Better installation support for various Linux distributions including
      Fedora 8 and Ubuntu.  The current runlevel is now acquired via the
      "runlevel" command instead of attempting to read /etc/inittab (which
      does not even exist on Ubuntu 7.10), and there are new command line
      arguments --init-dir, --init-name, and --runlevel to allow the init
      directory, init script name, and the runlevel to be manually specified
      on the install.pl command line.
    - Updated psad to automatically handle situations where the either the
      /var/log/psad/fwdata file or the /var/log/messages file (whichever
      syslog is writing iptables log messages to) gets rotated.  The
      filehandle is closed and reopened if the file shrinks or if the inode
      changes.  This strategy is borrowed from how the fwknop project deals

psadwatchd.pl
#!/usr/bin/perl -w
#
#########################################################################
#
# File: psadwatchd.pl
#
# Purpose: psadwatchd checks on an interval of every five seconds to make
#          sure that both kmsgsd and psad are running on the box.  If
#          either daemon has died, psadwatchd will restart it notify each
#          email address in @email_addresses that the daemon has been
#          restarted.
#
# Author: Michael Rash (mbr@cipherdyne.org)
#
# Credits:  (see the CREDITS file)
#
# Copyright (C) 1999-2006 Michael Rash (mbr@cipherdyne.org)
#
# License (GNU Public License):
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307
#    USA
#
#########################################################################
#
# $Id$
#

use POSIX qw(setsid);
use Getopt::Long 'GetOptions';
use Sys::Hostname 'hostname';
use strict;

### path to default psad library directory for psad perl modules
my $psad_lib_dir = '/usr/lib/psad';

### establish the default path to the config file (can be
### over-ridden with the -c <file> command line option.
my $config_file      = '/etc/psad/psadwatchd.conf';
my $psad_config_file = '/etc/psad/psad.conf';  ### for EMAIL_ADDRESSES

### default config file for ALERTING_METHODS keyword, which
#### is referenced by both psad and psadwatchd.  This keyword
#### allows email alerting or syslog alerting (or both) to be
#### disabled.
my $alerting_config_file = '/etc/psad/alert.conf';

my $warn_msg = '';
my $die_msg  = '';

### these vars are controled by the alert.conf file
my $no_email_alerts  = 0;
my $no_syslog_alerts = 0;

### configuration hash
my %config;

### commands hash
my %cmds;

### flag used for HUP signal
my $hup_flag = 0;

### handle command line arguments
die "[*] Specify the path to the psad.conf file with " .
    "\"-c <file>\".\n\n" unless (GetOptions (
    'config=s' => \$config_file
));

### import psad perl modules
&import_psad_perl_modules();

### import config
&import_config();

### make sure this is the only psadwatchd running on this system
&Psad::unique_pid($config{'PSADWATCHD_PID_FILE'});

### install WARN and DIE handlers
$SIG{'__WARN__'} = \&warn_handler;
$SIG{'__DIE__'}  = \&die_handler;

### install HUP handler so config can be re-imported
$SIG{'HUP'}  = \&hup_sig;

my $pid = fork;
exit if $pid;
die "[*] $0: Couldn't fork: $!" unless defined($pid);
POSIX::setsid() or die "[*] $0: Can't start a new session: $!\n";

### write the pid to the pid file
&Psad::writepid($config{'PSADWATCHD_PID_FILE'});

my $HOSTNAME = hostname;

### get the psad command line args
my $psad_Cmdline = &get_psad_Cmdline($config{'PSAD_CMDLINE_FILE'});

my ($d_emails, $k_emails, $p_emails) = (0,0,0);

#=================== end main ==================
### main loop
for (;;) {

    if ($hup_flag) {
        ### clear the HUP flag
        $hup_flag = 0;
        &import_config();
        &Psad::psyslog('psad(psadwatchd)', 'received HUP signal, ' .
            're-importing psadwatchd.conf') unless $no_syslog_alerts;
    }

    &check_process('psad', $psad_Cmdline,
        $config{'PSAD_PID_FILE'}, \$p_emails);
    &check_process('kmsgsd', '',
        $config{'KMSGSD_PID_FILE'}, \$k_emails);

    if ($die_msg) {
        &Psad::print_sys_msg($die_msg, "$config{'PSAD_DIR'}/errs/psadwatchd.die");
        $die_msg = '';
    }

    if ($warn_msg) {
        &Psad::print_sys_msg($warn_msg, "$config{'PSAD_DIR'}/errs/psadwatchd.warn");
        $warn_msg = '';
    }

    sleep $config{'PSADWATCHD_CHECK_INTERVAL'};
}
exit 0;
#=================== end main ==================

sub check_process() {
    my ($pidname, $pidcmdline, $pidfile, $email_count_ref) = @_;
    if (-e $pidfile) {
        unless (&Psad::pidrunning($pidfile)) {
            ### the daemon is not running so start it with $pidcmdline
            ### args (which may be empty)
            if ($$email_count_ref > $config{'PSADWATCHD_MAX_RETRIES'}) {
                ### this will exit the program
                &give_up($pidname);
            }
            ### should check the rv of this system() call
            system "$cmds{$pidname} $pidcmdline";
            my $subject = "[*] psadwatchd: restarted $pidname on $HOSTNAME";
            &Psad::sendmail($subject, '', $config{'EMAIL_ADDRESSES'},
                $cmds{'mail'}) unless $no_email_alerts;
            $$email_count_ref++;
            return;
        } else {
            ### the program is running now, so reset the watch count to zero
            $$email_count_ref = 0;
        }
    } else {
        my $subject = "[*] psadwatchd: pid file $pidfile\" does not exist " .
            "for $pidname.  Starting $pidname daemon.";
        &Psad::sendmail($subject, '', $config{'EMAIL_ADDRESSES'},
            $cmds{'mail'}) unless $no_email_alerts;
        ### start $pidname
        system "$cmds{$pidname} $pidcmdline";
    }
    return;
}

sub import_psad_perl_modules() {

    my $mod_paths_ar = &get_psad_mod_paths();

    push @$mod_paths_ar, @INC;
    splice @INC, 0, $#$mod_paths_ar+1, @$mod_paths_ar;

    require Psad;

    return;
}

sub get_psad_mod_paths() {

    my @paths = ();

    unless (-d $psad_lib_dir) {
        my $dir_tmp = $psad_lib_dir;
        $dir_tmp =~ s|lib/|lib64/|;
        if (-d $dir_tmp) {
            $psad_lib_dir = $dir_tmp;
        } else {
            die "[*] psad lib directory: $psad_lib_dir does not exist, ",
                "use --Lib-dir <dir>";
        }
    }

    opendir D, $psad_lib_dir or die "[*] Could not open $psad_lib_dir: $!";
    my @dirs = readdir D;
    closedir D;

    push @paths, $psad_lib_dir;

    for my $dir (@dirs) {
        ### get directories like "/usr/lib/psad/x86_64-linux"
        next unless -d "$psad_lib_dir/$dir";
        push @paths, "$psad_lib_dir/$dir"
            if $dir =~ m|linux| or $dir =~ m|thread|;
    }
    return \@paths;
}

sub get_psad_Cmdline() {
    my $psad_cmd_file = shift;
    my $noexit=0;
    my $psad_Cmdline;
    while ($noexit < 100) {
        if (-e $psad_cmd_file) {
            open CMD, "< $psad_cmd_file";
            $psad_Cmdline = <CMD>;
            close CMD;
            return $psad_Cmdline;  ### there may be _no_ command line args
        } else {
            $noexit++;
        }
        sleep 1;
    }
    my $subject = "[*] psadwatchd: psad is not running on $HOSTNAME.  " .
        "Please start it.";
    &Psad::sendmail($subject, '', $config{'EMAIL_ADDRESSES'}, $cmds{'mail'})
        unless $no_email_alerts;
    exit 0;
}

sub give_up() {
    my $pidname = shift;
    my $subject = "psadwatchd: restart limit reached for $pidname " .
                  "on $HOSTNAME!!!  Exiting.";
    &Psad::sendmail($subject, '', $config{'EMAIL_ADDRESSES'}, $cmds{'mail'})
        unless $no_email_alerts;
    exit 0;
}

sub import_config() {

    ### read in the configuration file
    &Psad::buildconf(\%config, \%cmds, $config_file);

    ### for EMAIL_ADDRESSES
    &Psad::buildconf(\%config, \%cmds, $psad_config_file);

    ### import alerting config (psadwatchd also references this file
    &Psad::buildconf(\%config, \%cmds, $alerting_config_file);

    ### expand any embedded vars within config values
    &Psad::expand_vars(\%config, \%cmds);

    ### make sure the configuration is complete
    &required_vars();

    $no_email_alerts = 1 if $config{'ALERTING_METHODS'} =~ /no.?e?mail/i;
    $no_syslog_alerts = 1 if $config{'ALERTING_METHODS'} =~ /no.?syslog/i;

    ### Check to make sure the commands specified in the config section
    ### are in the right place, and attempt to correct automatically if not.
    &Psad::check_commands(\%cmds, {'mail' => ''});

    return;
}

sub required_vars() {
    my @required_vars = qw(
        PSAD_PID_FILE PSAD_CMDLINE_FILE
        PSADWATCHD_PID_FILE EMAIL_ADDRESSES
        PSADWATCHD_CHECK_INTERVAL
        PSADWATCHD_MAX_RETRIES
        KMSGSD_PID_FILE
    );
    &Psad::defined_vars(\%config, $config_file, \@required_vars);
    return;
}

sub hup_sig() {
    $hup_flag = 1;
    return;
}

sub die_handler() {
    $die_msg = shift;
    return;
}

sub warn_handler() {
    $warn_msg = shift;
    return;
}
